ably/realtime_channel_spec_test.go

```diff
@@ -155,7 +155,804 @@ func attachAndWait(t *testing.T, channel *ably.RealtimeChannel) {
+		if expected, got := ably.ChannelStateFailed, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		err = channel.Detach(ctx)
+
+		if expected, got := "cannot Detach channel because it is in FAILED state", err.Error(); !strings.Contains(got, expected) {
+			t.Fatalf("expected error %+v; got %v", expected, got)
+		}
+		ablytest.Instantly.NoRecv(t, nil, channelStateChanges, t.Fatalf) // Should not make any change to the channel state
+
+	})
+
+	t.Run("RTL5d RTL5e: If connected, should do successful detach with server", func(t *testing.T) {
+		t.Parallel()
+
+		app, client := ablytest.NewRealtime(ably.WithAutoConnect(false))
+		defer safeclose(t, ablytest.FullRealtimeCloser(client), app)
+
+		connectionStateChanges := make(ably.ConnStateChanges, 10)
+		off := client.Connection.OnAll(connectionStateChanges.Receive)
+		defer off()
+
+		client.Connect()
+
+		var connectionChange ably.ConnectionStateChange
+
+		ablytest.Soon.Recv(t, &connectionChange, connectionStateChanges, t.Fatalf)
+		if expected, got := ably.ConnectionStateConnecting, connectionChange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, connectionChange)
+		}
+
+		ablytest.Soon.Recv(t, &connectionChange, connectionStateChanges, t.Fatalf)
+		if expected, got := ably.ConnectionStateConnected, connectionChange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, connectionChange)
+		}
+
+		channel := client.Channels.Get("test")
+		channelStateChanges := make(ably.ChannelStateChanges, 10)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		channeloff := channel.OnAll(channelStateChanges.Receive)
+		defer channeloff()
+
+		var channelStatechange ably.ChannelStateChange
+
+		err := channel.Attach(ctx)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		ablytest.Instantly.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateAttaching, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+
+		ablytest.Instantly.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateAttached, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+
+		err = channel.Detach(ctx)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		ablytest.Instantly.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateDetaching, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+
+		ablytest.Instantly.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
```

Comment:
```suggestion
		ablytest.Soon.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
```

`Instantly` really means that it should happen synchronously, with no wait except for CPU. This actually waits for Ably servers.

Reply:
Well, detach operation is already successful and it has already pushed the states in the buffered channel. You can check I have added nil check for err returned by detach which means detach will wait till channel is detached.

Reply:
Ah, that's true. Thanks.